packages/core/test/toPromise.test.ts

```diff
@@ -0,0 +1,287 @@
+          id: 'machine',
+          initial: 'pending',
+          after: {
+            800: 'done'
+          },
+          states: {
+            pending: {
+              invoke: {
+                src: 'fetch',
+                onError: {
+                  target: 'done'
+                }
+              }
+            },
+            done: {
+              type: 'final'
+            }
+          }
+        })
+      );
+
+      expect(errors).toEqual([]);
+    });
+
+    it('Should not allow you to self-target from the timeout', () => {
+      const errors = validatePromiseMachine(
+        createMachine({
+          id: 'machine',
+          initial: 'pending',
+
+          states: {
+            pending: {
+              invoke: {
+                src: 'fetch',
+                onError: {
+                  target: 'done'
+                }
+              },
+              after: {
+                800: 'pending'
+              }
+            },
+            done: {
+              type: 'final'
+            }
+          }
+        })
+      );
+
+      expect(errors).toEqual([
+        {
+          stateId: 'machine.pending',
+          type: 'timeout-targets-self'
+        }
+      ]);
+    });
+
+    it('Should require a timeout without a cond', () => {
+      const errors = validatePromiseMachine(
+        createMachine({
+          id: 'machine',
+          initial: 'pending',
+
+          states: {
+            pending: {
+              after: {
+                800: { target: 'done', cond: 'someCond' }
+              },
+              invoke: {
+                src: 'fetch',
+                onError: {
+                  target: 'done'
+                }
+              }
+            },
+            done: {
+              type: 'final'
+            }
+          }
+        })
+      );
+
+      expect(errors).toEqual([
+        {
+          stateId: 'machine.pending',
+          type: 'no-timeout-without-cond'
+        }
+      ]);
+    });
+  });
+
+  describe('toPromise', () => {
+    it('Should invoke the machine and run services', async () => {
+      const mockFunc = jest.fn();
+      const state = await toPromise(
+        createMachine({
+          initial: 'idle',
+          after: {
+            10000: {
+              target: 'timedOut'
+            }
+          },
+          states: {
+            idle: {
+              invoke: {
+                src: async () => {
+                  return mockFunc();
+                },
+                onDone: {
+                  target: 'final'
+                },
+                onError: {
+                  target: 'errored'
+                }
+              }
+            },
+            timedOut: {
+              type: 'final'
+            },
+            errored: {
+              type: 'final'
+            },
+            final: {
+              type: 'final'
+            }
+          }
+        })
+      );
+
+      expect(state.value).toEqual('final');
```

Comment:
Separate discussion/PR: in the future, this would be a good opportunity to represent "done data" in the `state` itself (`undefined` by default). That way, we can use promise machines like this:

```js
const { doneData } = await toPromise(machine);

return doneData;
```

Reply:
Agree!